src/library/scala/collection/immutable/NVector.scala

```diff
@@ -0,0 +1,2004 @@
+  override def foreach[U](f: A => U): Unit = {
+    foreachElem(prefix1, f)
+    foreachElem(prefix2, f)
+    foreachElem(data3, f)
+    foreachElem(suffix2, f)
+    foreachElem(suffix1, f)
+  }
+
+  override def map[B](f: A => B): NVector[B] =
+    copy(prefix1 = mapElems1(prefix1, f), prefix2 = mapElems(2, prefix2, f),
+      data3 = mapElems(3, data3, f),
+      suffix2 = mapElems(2, suffix2, f), suffix1 = mapElems1(suffix1, f))
+
+  override def last: A = suffix1(suffix1.length-1).asInstanceOf[A]
+
+  protected[this] def slice0(lo: Int, hi: Int): NVector[A] = {
+    val b = new NVectorSliceBuilder(lo, hi)
+    b.consider(1, prefix1)
+    b.consider(2, prefix2)
+    b.consider(3, data3)
+    b.consider(2, suffix2)
+    b.consider(1, suffix1)
+    b.result()
+  }
+
+  override def tail: NVector[A] =
+    if(len1 > 1) copy(prefix1 = copyTail(prefix1), len1 = len1-1, len12 = len12-1, length = length-1)
+    else slice0(1, length)
+
+  override def init: NVector[A] =
+    if(suffix1.length > 1) copy(suffix1 = copyInit(suffix1), length = length-1)
+    else slice0(0, length-1)
+
+  protected[immutable] def vectorSliceCount: Int = 5
+  protected[immutable] def vectorSlice(idx: Int): Array[_ <: AnyRef] = (idx: @switch) match {
+    case 0 => prefix1
+    case 1 => prefix2
+    case 2 => data3
+    case 3 => suffix2
+    case 4 => suffix1
+  }
+  protected[immutable] def vectorSlicePrefixLength(idx: Int): Int = (idx: @switch) match {
+    case 0 => len1
+    case 1 => len12
+    case 2 => len12 + data3.length*WIDTH2
+    case 3 => length - suffix1.length
+    case 4 => length
+  }
+
+  override protected[this] def appendedAll0[B >: A](suffix: collection.IterableOnce[B], k: Int): NVector[B] = {
+    val suffix1b = append1IfSpace(suffix1, suffix)
+    if(suffix1b ne null) copy(suffix1 = suffix1b, length = length-suffix1.length+suffix1b.length)
+    else super.appendedAll0(suffix, k)
+  }
+}
+
+
+/** 4-dimensional radix-balanced finger tree */
+private final class NVector4[+A](_prefix1: Arr1, len1: Int,
+                                 prefix2: Arr2, len12: Int,
+                                 prefix3: Arr3, len123: Int,
+                                 data4: Arr4,
+                                 suffix3: Arr3, suffix2: Arr2, suffix1: Arr1,
+                                 val length: Int) extends NVector[A](_prefix1) {
+  import NVectorStatics._
+
+  @inline private[this] def copy(prefix1: Arr1 = prefix1, len1: Int = len1,
+                                 prefix2: Arr2 = prefix2, len12: Int = len12,
+                                 prefix3: Arr3 = prefix3, len123: Int = len123,
+                                 data4: Arr4 = data4,
+                                 suffix3: Arr3 = suffix3, suffix2: Arr2 = suffix2, suffix1: Arr1 = suffix1,
+                                 length: Int = length) =
+    new NVector4(prefix1, len1, prefix2, len12, prefix3, len123, data4, suffix3, suffix2, suffix1, length)
+
+  @inline def apply(index: Int): A = {
```

Comment:
bytecodesize-wise this is too large to meaningfully inline by the JIT, is there any way of pruning it or splitting it up into many, small, private-final invocations?

Reply:
It won't get inlined anyway due to the usually megamorphic call site.